src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,171 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> nextNode;
+        private Node<T> previousNode;
+
+        public Node(Node<T> previousNode, T value, Node<T> nextNode) {
+            this.nextNode = nextNode;
+            this.value = value;
+            this.previousNode = previousNode;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            Node<T> newNode = new Node<>(null, value, null);
+            head = newNode;
+            tail = newNode;
+        } else {
+            Node<T> newNode = new Node<>(tail, null, null);
+            tail.nextNode = newNode;
+            tail = newNode;
+            newNode.value = value;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexCheckForAdd(index);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> insertingNode = new Node<>(null, value, head);
+            head = insertingNode;
+            size++;
+        } else {
+            Node<T> insertingNode;
+            Node newNode = head;
+            for (int i = 0; i < index - 1; i++) {
+                newNode = newNode.nextNode;
+            }
+            insertingNode = new Node<>(newNode, value, newNode.nextNode);
+            Node afteradditingNode = newNode.nextNode;
+            newNode.nextNode = insertingNode;
+            insertingNode.previousNode = newNode;
+            afteradditingNode.previousNode = insertingNode;
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int j = 0; j < list.size(); j++) {
+            add((T) list.get(j));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        Node getteadNote = whatHalf(index);
+        return (T) getteadNote.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        Node<T> currentNode = whatHalf(index);
+        T valueReplacing = (T) currentNode.value;
+        currentNode.value = value;
+        return valueReplacing;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        Node<T> currentNode = whatHalf(index);
+        if (size / 2 - 1 < index) {
```

Comment:
why do we need this checking here?

Reply:
deleting formula :
currentNode.nextNode = currentNode.previousNode;
or
currentNode.previousNode = currentNode.nextNode;
depends on my way of iteration: descend or ascend